### PR TITLE
Refactor chains/nsmgr to use Option pattern

### LINKED ISSUE #733: Refactor chains/nsmgr to use Option pattern
## Motivation

We can simplify using of our chains with
[Option pattern](https://github.com/tmrts/go-patterns/blob/master/idiom/functional-options.md)

See at an example: https://github.com/networkservicemesh/sdk/commit/9007c4530373d301993ecd8985a1c9abebe359b2

Moreover, this can super help us with minimization of manual changes of upstream repositories.

Entry point: https://github.com/networkservicemesh/sdk/blob/master/pkg/networkservice/chains/nsmgr/server.go

### PR DESCRIPTION
**Pull request by haiodo**
Fixes #733 
Refactor chains/nsmgr to use Option pattern

Signed-off-by: Andrey Sobolev <haiodo@gmail.com>

### CODE REVIEW

File: pkg/networkservice/chains/nsmgr/server.go [hunk 1]
```diff
@@ -69,20 +70,76 @@ type nsmgrServer struct {
...
+// WithName - set a nsmgr name, a default name is `Nsmgr`.
+func WithName(name string) Option {
+	return func(o *serverOptions) {
+		o.name = name
+	}
+}
+
+// WithURL - set a public URL address for NetworkServiceManager, it is used to access endpoints within this NSMgr from remote clients.
+// Default value is not set, and endpoints address will not be changed.
+func WithURL(url string) Option {
```
Comment (denis-tingaikin) [inline pkg/networkservice/chains/nsmgr/server.go:117]: URL should be required parameter.

File: pkg/networkservice/chains/nsmgr/server.go [hunk 2]
```diff
@@ -69,20 +70,76 @@ type nsmgrServer struct {
...
-//           authzServer - authorization server chain element
 //           tokenGenerator - authorization token generator
-//           registryCC - client connection to reach the upstream registry, could be nil, in this case only in memory storage will be used.
-// 			 clientDialOptions -  a grpc.DialOption's to be passed to GRPC connections.
-func NewServer(ctx context.Context, nsmRegistration *registryapi.NetworkServiceEndpoint, authzServer networkservice.NetworkServiceServer, tokenGenerator token.GeneratorFunc, registryCC grpc.ClientConnInterface, clientDialOptions ...grpc.DialOption) Nsmgr {
+//			 options - a set of NSMGR options.
+func NewServer(ctx context.Context, tokenGenerator token.GeneratorFunc, options ...Option) Nsmgr {
+	opts := &serverOptions{
+		authorizeServer: authorize.NewServer(authorize.Any()),
+		name:            "Nsmgr",
```
Comment (denis-tingaikin) [inline pkg/networkservice/chains/nsmgr/server.go:131]:
```suggestion
		name:            "nsmgr-"+uuid.New().String(),
```

**Review by denis-tingaikin — CHANGES_REQUESTED** [review]

File: pkg/networkservice/chains/nsmgr/server.go (hunk 1 above)
Comment (haiodo) [reply pkg/networkservice/chains/nsmgr/server.go:117]: It is not required for Local -> Nsmgr -> local only cases and for cases if we doesn't use public Url for nsmgr.

Comment (denis-tingaikin) [reply pkg/networkservice/chains/nsmgr/server.go:117]: OK, currently we have not a test for this scenario actually. Let's then cover it :)

File: pkg/networkservice/chains/nsmgr/server.go (hunk 2 above)
Comment (haiodo) [reply pkg/networkservice/chains/nsmgr/server.go:131]: Thanks.

File: pkg/networkservice/chains/nsmgr/server.go (hunk 1 above)
Comment (haiodo) [reply pkg/networkservice/chains/nsmgr/server.go:121]: Added test with a Unix Socket nsmgr and empty Url.

**Comment by denis-tingaikin** [comment]
Note: For some reason new test produces goleaks: https://github.com/networkservicemesh/sdk/pull/795/checks?check_run_id=2218240245

File: pkg/tools/sandbox/builder.go [hunk 3]
```diff
@@ -269,20 +300,37 @@ func (b *Builder) newNSMgr(ctx context.Context, address string, registryURL *url
...
+	if b.useUnixSockets {
+		serveURL = grpcutils.TargetToURL(address)
+	} else {
+		listener, err := net.Listen("tcp", address)
+		b.require.NoError(err)
+		serveURL = grpcutils.AddressToURL(listener.Addr())
+		b.require.NoError(listener.Close())
+	}
+
+	nsmgrName := "nsmgr-" + uuid.New().String()
```
Comment (denis-tingaikin) [inline pkg/tools/sandbox/builder.go:314]: let's use default value
```suggestion
```

File: pkg/tools/sandbox/builder.go [hunk 4]
```diff
@@ -360,6 +409,16 @@ func (b *Builder) SetupRegistryClients(ctx context.Context, node *Node) {
...
 }
 
+// newAddress - will return a new public address, if unixSockets are used prefix will be used to make uniq files.
+func (b *Builder) newAddress(prefix string) string {
+	if !b.useUnixSockets {
+		return "127.0.0.1:0"
+	}
+
+	b.addressID++
+	return fmt.Sprintf("unix:%s/%s_%d.sock", b.sockPath, prefix, b.addressID)
```
Comment (denis-tingaikin) [inline pkg/tools/sandbox/builder.go:419]: Can we use uuid instead of `b.addressID`?

Comment (haiodo) [reply pkg/tools/sandbox/builder.go]: Sure :-) Fixed.

**Comment by haiodo** [comment]
@denis-tingaikin Unix sockets with uuid id inside are not working, so I've revert a logic to use an index, and since index are inside random temporary folder it should be fully safe :-)

**Review by denis-tingaikin — APPROVED** [review]

**Pull request merged**